File: LLD/Design_Pattern/Structural_Patterns/cpp/2_decorator_pattern/pizzaExample.cpp
#include <iostream>
using namespace std;

class BasePizza
{
public:
    virtual int cost() = 0;
};

class FarmHouse : public BasePizza
{
public:
    int cost()
    {
        return 300;
    }
};

class VegDelight : public BasePizza
{
public:
    int cost()
    {
        return 250;
    }
};

// Decorators

class ToppingDecorator : public BasePizza
{
protected:
    BasePizza *pizza;
};

class ExtraCheese : public ToppingDecorator
{
public:
    ExtraCheese(BasePizza *pizza)
    {
        this->pizza = pizza;
    }
    int cost()
    {
        return this->pizza->cost() + 40;
    }
};

class Mushroom : public ToppingDecorator
{
public:
    Mushroom(BasePizza *pizza)
    {
        this->pizza = pizza;
    }
    int cost()
    {
        return this->pizza->cost() + 80;
    }
};

void printToppingsMenu()
{
    cout << "Choose Toppings \n";
    cout << "1. Extracheese\n";
    cout << "2. Mushroom\n";
    cout << "Press any other key to exit\n";
    cout << "Your choice : ";
}

int main()
{
    // BasePizza *bp = new ExtraCheese(new FarmHouse());
    // cout << "Farmhouse cost :" << bp->cost() << endl;

    BasePizza *b = nullptr;

    int option;
    cout << "Select pizza type \n1.Farmhouse\t2.VegDelight\t3.Margherita\n";
    cin >> option;

    switch (option)
    {
    case 1:
        b = new FarmHouse();
        break;
    case 2:
        b = new VegDelight();
        break;

    default:
        break;
    }

    bool more = true;
    do
    {
        cout << "Current Total : " << b->cost() << endl;
        printToppingsMenu();
        cin >> option;

        switch (option)
        {
        case 1:
            b = new ExtraCheese(b);
            break;
        case 2:
            b = new Mushroom(b);
            break;
        default:
            more = false;
            break;
        }

    } while (more);

    cout << "Total cost : " << b->cost() << endl;
}

File: LLD/Design_Pattern/Structural_Patterns/typescript/2_decorator_pattern/pizzaExample.ts
/*
Problem Statement : 


*/

abstract class BasePizza{
    public abstract cost(): number;
}

class FarmHouse extends BasePizza{
    public cost(): number {
        return 200;
    }
}

class VegDelight extends BasePizza{
    
    public cost(): number {
        return 300;
    }
}

class Margherita extends BasePizza{
    public cost(): number {
        return 350;
    }
}


//Decorators
abstract class ToppingDecorator extends BasePizza{
    
}

class ExtraCheese extends ToppingDecorator{
    pizza: BasePizza;
    constructor(pizza: BasePizza) {
        super();
        this.pizza = pizza;
    }

    public cost(): number {
        return this.pizza.cost() +30;
    }
}

class Mushroom extends ToppingDecorator{
    pizza: BasePizza;
    constructor(pizza: BasePizza) {
        super()
        this.pizza = pizza;
    }
    public cost(): number {
        return this.pizza.cost() + 50;
    }
}


const p1 = new FarmHouse();
console.log("Price of FarmHouse : ",p1.cost());


const p2 = new ExtraCheese(new FarmHouse());
console.log("Price of FarmHouse pizza with extracheese :", p2.cost());

const p3 = new Mushroom(new ExtraCheese(new FarmHouse()));
console.log("Price of Farmhouse pizza with extracheese and mushroom topping : ", p3.cost());


// take user input

// import * as readline from 'readline';
// import { stdin as input, stdout as output } from 'process'
// const rl = readline.createInterface({ input, output })

// console.log("Base pizza \n1.FarmHouse\n2.VegDelight\n3.Margherita");


// function showToppingsMenu(){
//     return ("Toppings: \n1.Extracheese \n2.Mushroom\nPress any other key to exit");
// }

// let myPizza: BasePizza;
// rl.question("Base pizza", (bp) => {
//     switch (bp) {
//         case "1":
//             myPizza = new FarmHouse();
//             break;
//         case "2":
//             myPizza = new VegDelight();
//             break;
//         case "3":
//             myPizza = new Margherita();
//             break;
//         default:
//             myPizza = new FarmHouse();
//             break;
//     }

//     while (true) {
//         let myOption: string = "";
//         rl.question(showToppingsMenu(), (option) => {
//             myOption = option;
//             switch (option) {
//                 case "1":
//                     myPizza = new ExtraCheese(myPizza);
//                     break;
//                 case "2":
//                     myPizza = new Mushroom(myPizza);
//                     break;
//                 default:
//                     break;
//             }

            
//         })
//         if (!["1", "2"].includes(myOption)) {
//             console.log('exiting');
//             break;
//         }
        
//     }

//     rl.close();
// })